File: Snake_PyQt/Snake.py
#蛇类
class Snake:
    #初始化函数，传入头部坐标，节数和初始方向初始化蛇
    def __init__(self, head, sectionNum, direction):
        #蛇的方向改变时的头部按方向变化
        self.dirChange = [(0, -1), (0, 1), (-1, 0), (1, 0)]

        #初始化初始坐标
        self.dir = direction

        #蛇身坐标
        self.pos = [head]
        for i in range(sectionNum):
            last = self.pos[len(self.pos) - 1]
            self.pos.append([
                last[0] - self.dirChange[self.dir][0],
                last[1] - self.dirChange[self.dir][1]
            ])

    #移动
    def move(self):
        #把原来的尾巴弹出来
        self.pos.pop()
        #获取老头
        head = self.pos[0]
        #把新头压入队列头，作为真正的新头，并压入蛇身数组
        self.pos.insert(0, [
            head[0] + self.dirChange[self.dir][0],
            head[1] + self.dirChange[self.dir][1]
        ])

    #是否死亡
    def isDead(self, up, down, left, right):
        head = self.pos[0]
        #撞墙
        if head[0] < left or head[0] > right or head[1] < up or head[1] > down:
            return True
        #撞到蛇身
        else:
            for i in range(1, len(self.pos)):
                if head[0] == self.pos[i][0] and head[1] == self.pos[i][1]:
                    return True
            return False

    #吃到食物
    def eatFood(self, food):
        head = self.pos[0]
        #如果吃到食物
        if head[0] == food.x and head[1] == food.y:
            #按照相应的节数进行增长
            for i in range(food.bodyGrowth):
                last = self.pos[len(self.pos) - 1]
                secondLast = self.pos[len(self.pos) - 2]
                self.pos.append([
                    last[0] + last[0] - secondLast[0],
                    last[1] + last[1] - secondLast[1]
                ])
            return True
        else:
            return False

File: snake_qt5/Snake.cpp
#include "Snake.h"

Snake::Snake() :Snake({}, 0, 3)
{

}

Snake::Snake(QPair<int, int> head, int sectionNum, int nowDir) : pos({}), dirChange({ { 0,-1 },{ 0,1 },{ -1,0 },{ 1,0 } }), nowDir(Dir::right)
{
	//确定当前方向
	nowDir = (nowDir >= 0 ? nowDir : -nowDir) % 4;
	this->nowDir = static_cast<Dir>(nowDir);

	//初始化蛇身坐标
	pos.append(head);
	for (int i = 0; i < sectionNum - 1; i++)
	{
		QPair<int, int> last = pos.last();
		last.first -= dirChange[this->nowDir].first;
		last.second -= dirChange[this->nowDir].second;
		pos.append(last);
	}
}

Snake::~Snake()
{

}

const QVector<QPair<int, int>> Snake::getPos() const
{
	return pos;
}

void Snake::move()
{
	pos.pop_back(); //推出最后一个
	QPair<int, int> newHead(pos.at(0)); //取头用来构造新头
	//把相应的xy的变化加到新头上
	newHead.first += dirChange[nowDir].first;
	newHead.second += dirChange[nowDir].second;
	//把新头从头部压入蛇身坐标数组，作为真正的新头
	pos.push_front(newHead);
}

bool Snake::isDead(int up, int down, int left, int right)
{
	//判定是否撞到墙
	if (pos[0].first<left || pos[0].first>right || pos[0].second<up || pos[0].second>down)
	{
		return true;
	}
	//否则判定是否自身相撞
	else
	{
		for (int i = 1; i < pos.size(); i++)
		{
			if (pos[0].first == pos[i].first&&pos[0].second == pos[i].second)
			{
				return true;
			}
		}
		return false;
	}
}

void Snake::setNowDir(int nowDir)
{
	this->nowDir = static_cast<Dir>(nowDir % 4);
}

bool Snake::eatFood(Food food)
{
	if (pos[0].first == food.getPos().first&&pos[0].second == food.getPos().second)
	{
		//吃一次食物增长相应的节数
		for (int i = 0; i < food.getBodyGrowth(); i++)
		{
			QPair<int, int> last = pos[pos.size() - 1]; //当前最后一节
			QPair<int, int> secondLast = pos[pos.size() - 2]; //当前倒数第二节
			QPair<int, int> newLast = last; //新的最后一节
			newLast.first += last.first - secondLast.first;
			newLast.second += last.second - secondLast.second;
			pos.append(newLast);
		}
		return true;
	}
	else
	{
		return false;
	}
}